File: frontend/src/types/dto/budget.ts
// DTO mirrors of backend/app/schemas/budget.py (§12.4 срез-1/срез-2, budget feature flag).

export type BudgetDomain = "training" | "medical" | "events";

/** overview дополнительно возвращает read-only "ppe" псевдо-домен */
export type BudgetOverviewDomainCode = BudgetDomain | "ppe";

export type BreakdownDimension =
  | "article"
  | "domain"
  | "company"
  | "branch"
  | "site";

export interface SafetyBudgetDto {
  id: string;
  name: string;
  domain: BudgetDomain;
  period_start: string;
  period_end: string;
  planned_amount: number;
  notes: string | null;
}

export interface SafetyBudgetPageDto {
  items: SafetyBudgetDto[];
  total: number;
  limit: number;
  offset: number;
}

export interface BudgetArticleActualDto {
  article_id: string | null;
  /** "— без статьи" для NULL article_id */
  article_name: string;
  amount: number;
}

export interface SafetyBudgetDetailDto extends SafetyBudgetDto {
  actual_total: number;
  /** planned - actual; отрицательный = перерасход */
  remaining: number;
  expense_count: number;
  by_article: BudgetArticleActualDto[];
}

export interface SafetyBudgetCreateInput {
  name: string;
  domain: BudgetDomain;
  period_start: string;
  period_end: string;
  planned_amount: number;
  notes?: string | null;
}

export interface SafetyBudgetUpdateInput {
  name?: string;
  period_start?: string;
  period_end?: string;
  planned_amount?: number;
  notes?: string | null;
}

export interface BudgetArticleDto {
  id: string;
  code: string;
  name: string;
  /** NULL = универсальная статья (не привязана к домену) */
  domain: BudgetDomain | null;
  is_active: boolean;
}

export interface BudgetArticlePageDto {
  items: BudgetArticleDto[];
  total: number;
  limit: number;
  offset: number;
}

export interface BudgetArticleCreateInput {
  code: string;
  name: string;
  domain?: BudgetDomain | null;
  is_active?: boolean;
}

export interface BudgetArticleUpdateInput {
  name?: string;
  domain?: BudgetDomain | null;
  is_active?: boolean;
  // code иммутабелен (ключ уникальности/сида)
}

export interface BudgetSeedResultDto {
  created: number;
  skipped: number;
}

export interface BudgetExpenseDto {
  id: string;
  domain: BudgetDomain;
  article_id: string | null;
  /** outerjoin в сервисе; null если article_id === null */
  article_name: string | null;
  title: string;
  occurred_on: string;
  amount: number;
  company_id: string | null;
  branch_id: string | null;
  site_id: string | null;
  entity_type: string | null;
  entity_id: string | null;
  notes: string | null;
}

export interface BudgetExpensePageDto {
  items: BudgetExpenseDto[];
  total: number;
  limit: number;
  offset: number;
}

export interface BudgetExpenseCreateInput {
  domain: BudgetDomain;
  article_id?: string | null;
  title: string;
  occurred_on: string;
  amount: number;
  company_id?: string | null;
  branch_id?: string | null;
  site_id?: string | null;
  entity_type?: string | null;
  entity_id?: string | null;
  notes?: string | null;
}

export interface BudgetExpenseUpdateInput {
  article_id?: string | null;
  title?: string;
  occurred_on?: string;
  amount?: number;
  company_id?: string | null;
  branch_id?: string | null;
  site_id?: string | null;
  entity_type?: string | null;
  entity_id?: string | null;
  notes?: string | null;
  // domain иммутабелен (перенос расхода между доменами = удалить+создать)
}

export interface BudgetOverviewBudgetRowDto {
  id: string;
  name: string;
  period_start: string;
  period_end: string;
  planned_amount: number;
  actual_own_period: number;
  remaining: number;
}

export interface BudgetOverviewDomainDto {
  domain: BudgetOverviewDomainCode;
  /** True только для ppe */
  read_only: boolean;
  planned: number;
  actual: number;
  remaining: number;
  /** только ppe */
  warning_unpriced_receipts?: number | null;
  budgets: BudgetOverviewBudgetRowDto[];
}

export interface BudgetOverviewDto {
  generated_at: string;
  date_from: string;
  date_to: string;
  domains: BudgetOverviewDomainDto[];
}

export interface BudgetBreakdownItemDto {
  /** "" = None-bucket «— без привязки»/«— без статьи» */
  id: string;
  name: string;
  amount: number;
}

export interface BudgetBreakdownDto {
  dimension: BreakdownDimension;
  date_from: string;
  date_to: string;
  /** строк ДО cap 200 */
  total: number;
  items: BudgetBreakdownItemDto[];
}

/** FSM: draft -> submitted -> approved -> paid; submitted -> rejected. rejected/paid терминальны. */
export type BudgetReimbursementStatus =
  | "draft"
  | "submitted"
  | "approved"
  | "rejected"
  | "paid";

export interface BudgetReimbursementDto {
  id: string;
  title: string;
  status: BudgetReimbursementStatus;
  period_start: string;
  period_end: string;
  requested_amount: number;
  /** заполняется на approve; по умолчанию = requested_amount */
  approved_amount: number | null;
  /** внешний номер заявки в СФР */
  reference: string | null;
  company_id: string | null;
  /** обязательна при reject */
  decision_reason: string | null;
  submitted_at: string | null;
  decided_at: string | null;
  paid_at: string | null;
  notes: string | null;
  item_count: number;
  /** сумма связанных расходов (может расходиться с requested_amount) */
  items_amount: number;
}

export interface BudgetReimbursementItemDto {
  expense_id: string;
  title: string;
  domain: BudgetDomain;
  occurred_on: string;
  amount: number;
}

export interface BudgetReimbursementDetailDto extends BudgetReimbursementDto {
  items: BudgetReimbursementItemDto[];
}

export interface BudgetReimbursementPageDto {
  items: BudgetReimbursementDto[];
  total: number;
  limit: number;
  offset: number;
}

export interface BudgetReimbursementCreateInput {
  title: string;
  period_start: string;
  period_end: string;
  requested_amount: number;
  reference?: string | null;
  company_id?: string | null;
  notes?: string | null;
  // status не передаётся: заявка всегда создаётся в draft
}

export interface BudgetReimbursementUpdateInput {
  title?: string;
  period_start?: string;
  period_end?: string;
  requested_amount?: number;
  reference?: string | null;
  company_id?: string | null;
  notes?: string | null;
  // редактировать можно только draft (иначе 409); статус меняется только через действия FSM
}

File: backend/app/schemas/budget.py
"""Схемы бюджетного контура §12.4 (срез-1)."""

from __future__ import annotations

from datetime import date, datetime

from pydantic import Field, model_validator

from app.models.budget import BUDGET_DOMAINS
from app.schemas.base import BaseSchema


def _check_period(values):
    if values.period_end < values.period_start:
        raise ValueError("period_end must be >= period_start")
    return values


class SafetyBudgetCreate(BaseSchema):
    name: str = Field(min_length=1, max_length=255)
    domain: str
    period_start: date
    period_end: date
    planned_amount: float = Field(ge=0)
    notes: str | None = Field(default=None, max_length=1000)

    @model_validator(mode="after")
    def _validate(self):
        if self.domain not in BUDGET_DOMAINS:
            raise ValueError(f"unknown domain: {self.domain}")
        return _check_period(self)


class SafetyBudgetUpdate(BaseSchema):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    period_start: date | None = None
    period_end: date | None = None
    planned_amount: float | None = Field(default=None, ge=0)
    notes: str | None = Field(default=None, max_length=1000)
    # domain намеренно НЕ обновляется (перенос бюджета между доменами = удалить+создать)


class SafetyBudgetRead(BaseSchema):
    id: str
    name: str
    domain: str
    period_start: date
    period_end: date
    planned_amount: float
    notes: str | None


class SafetyBudgetPage(BaseSchema):
    items: list[SafetyBudgetRead]
    total: int
    limit: int
    offset: int


class BudgetArticleActualRead(BaseSchema):
    article_id: str | None
    article_name: str  # "— без статьи" для NULL
    amount: float


class SafetyBudgetDetail(SafetyBudgetRead):
    actual_total: float
    remaining: float  # planned - actual; отрицательный = перерасход
    expense_count: int
    by_article: list[BudgetArticleActualRead]


class BudgetArticleCreate(BaseSchema):
    code: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=255)
    domain: str | None = None  # NULL = универсальная
    is_active: bool = True

    @model_validator(mode="after")
    def _validate(self):
        if self.domain is not None and self.domain not in BUDGET_DOMAINS:
            raise ValueError(f"unknown domain: {self.domain}")
        return self


class BudgetArticleUpdate(BaseSchema):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    domain: str | None = None
    is_active: bool | None = None
    # code иммутабелен (ключ уникальности/сида)

    @model_validator(mode="after")
    def _validate(self):
        if self.domain is not None and self.domain not in BUDGET_DOMAINS:
            raise ValueError(f"unknown domain: {self.domain}")
        return self


class BudgetArticleRead(BaseSchema):
    id: str
    code: str
    name: str
    domain: str | None
    is_active: bool


class BudgetArticlePage(BaseSchema):
    items: list[BudgetArticleRead]
    total: int
    limit: int
    offset: int


class BudgetSeedResult(BaseSchema):
    created: int
    skipped: int


class BudgetExpenseCreate(BaseSchema):
    domain: str
    article_id: str | None = None
    title: str = Field(min_length=1, max_length=255)
    occurred_on: date
    amount: float = Field(gt=0)
    company_id: str | None = None
    branch_id: str | None = None
    site_id: str | None = None
    entity_type: str | None = None
    entity_id: str | None = None
    notes: str | None = Field(default=None, max_length=1000)

    @model_validator(mode="after")
    def _validate(self):
        if self.domain not in BUDGET_DOMAINS:
            raise ValueError(f"unknown domain: {self.domain}")
        if (self.entity_id is None) != (self.entity_type is None):
            raise ValueError("entity_type and entity_id must be provided together")
        return self


class BudgetExpenseUpdate(BaseSchema):
    article_id: str | None = None
    title: str | None = Field(default=None, min_length=1, max_length=255)
    occurred_on: date | None = None
    amount: float | None = Field(default=None, gt=0)
    company_id: str | None = None
    branch_id: str | None = None
    site_id: str | None = None
    entity_type: str | None = None
    entity_id: str | None = None
    notes: str | None = Field(default=None, max_length=1000)
    # domain иммутабелен (перенос расхода между доменами = удалить+создать)


class BudgetExpenseRead(BaseSchema):
    id: str
    domain: str
    article_id: str | None
    article_name: str | None  # outerjoin в сервисе
    title: str
    occurred_on: date
    amount: float
    company_id: str | None
    branch_id: str | None
    site_id: str | None
    entity_type: str | None
    entity_id: str | None
    notes: str | None


class BudgetExpensePage(BaseSchema):
    items: list[BudgetExpenseRead]
    total: int
    limit: int
    offset: int


class BudgetOverviewBudgetRow(BaseSchema):
    id: str
    name: str
    period_start: date
    period_end: date
    planned_amount: float
    actual_own_period: float
    remaining: float


class BudgetOverviewDomain(BaseSchema):
    domain: str  # training|medical|events|ppe
    read_only: bool  # True только для ppe
    planned: float
    actual: float
    remaining: float
    warning_unpriced_receipts: int | None = None  # только ppe
    budgets: list[BudgetOverviewBudgetRow]


class BudgetOverviewResponse(BaseSchema):
    generated_at: datetime
    date_from: date
    date_to: date
    domains: list[BudgetOverviewDomain]


class BudgetBreakdownItem(BaseSchema):
    id: str  # "" = None-bucket «— без привязки»/«— без статьи»
    name: str
    amount: float


class BudgetBreakdownResponse(BaseSchema):
    dimension: str
    date_from: date
    date_to: date
    total: int  # строк ДО cap 200
    items: list[BudgetBreakdownItem]
